File: backend/internal/server/routes/agent_management.go
package routes

import (
	"github.com/Wei-Shaw/sub2api/internal/handler"

	"github.com/gin-gonic/gin"
)

func RegisterAgentManagementRoutes(authenticated *gin.RouterGroup, h *handler.Handlers) {
	agentManagement := authenticated.Group("/agent-management")
	{
		agentManagement.GET("/summary", h.AgentManagement.Summary)
		agentManagement.GET("/direct-users", h.AgentManagement.ListDirectUsers)
		agentManagement.POST("/direct-users", h.AgentManagement.CreateDirectUser)
		agentManagement.GET("/direct-users/groups/assigned", h.AgentManagement.ListDirectUsersWithGroup)
		agentManagement.GET("/direct-users/groups/unassigned", h.AgentManagement.ListDirectUsersWithoutGroup)
		agentManagement.PUT("/direct-users/groups/batch", h.AgentManagement.SetDirectUsersGroupDelegationsBatch)
		agentManagement.PUT("/direct-users/groups/existing", h.AgentManagement.UpdateDirectUsersExistingGroupDelegations)
		agentManagement.POST("/direct-users/groups/reclaim", h.AgentManagement.ReclaimDirectUsersGroupDelegations)
		agentManagement.GET("/direct-agents", h.AgentManagement.ListDirectAgents)
		agentManagement.GET("/direct-agents/groups/assigned", h.AgentManagement.ListDirectAgentsWithGroup)
		agentManagement.GET("/direct-agents/groups/unassigned", h.AgentManagement.ListDirectAgentsWithoutGroup)
		agentManagement.PUT("/direct-agents/groups/batch", h.AgentManagement.SetDirectAgentsGroupDelegationsBatch)
		agentManagement.PUT("/direct-agents/groups/existing", h.AgentManagement.UpdateDirectAgentsExistingGroupDelegations)
		agentManagement.POST("/direct-agents/groups/reclaim", h.AgentManagement.ReclaimDirectAgentsGroupDelegations)
		agentManagement.GET("/direct-enterprises", h.AgentManagement.ListDirectEnterprises)
		agentManagement.GET("/direct-enterprises/groups/assigned", h.AgentManagement.ListDirectEnterprisesWithGroup)
		agentManagement.GET("/direct-enterprises/groups/unassigned", h.AgentManagement.ListDirectEnterprisesWithoutGroup)
		agentManagement.PUT("/direct-enterprises/groups/batch", h.AgentManagement.SetDirectEnterprisesGroupDelegationsBatch)
		agentManagement.PUT("/direct-enterprises/groups/existing", h.AgentManagement.UpdateDirectEnterprisesExistingGroupDelegations)
		agentManagement.POST("/direct-enterprises/groups/reclaim", h.AgentManagement.ReclaimDirectEnterprisesGroupDelegations)
		agentManagement.GET("/admin-agent-tree", h.AgentManagement.AdminAgentTree)
		agentManagement.GET("/structure", h.AgentManagement.SubordinateStructure)
		agentManagement.GET("/usage", h.AgentManagement.ListUsage)
		agentManagement.GET("/usage/stats", h.AgentManagement.UsageStats)
		agentManagement.GET("/usage/users", h.AgentManagement.ListUsageUsers)
		agentManagement.PUT("/children/:id/allocation", h.AgentManagement.UpdateAllocation)
		agentManagement.PUT("/children/:id/agent-income", h.AgentManagement.SetAgentIncome)
		agentManagement.PUT("/invite-defaults", h.AgentManagement.UpdateInviteDefaults)
		agentManagement.POST("/children/:id/upgrade", h.AgentManagement.UpgradeDirectUser)
		agentManagement.DELETE("/children/:id", h.AgentManagement.DeleteDirectChild)
		agentManagement.GET("/groups", h.AgentManagement.ListMyGroups)
		agentManagement.GET("/invite-default-groups", h.AgentManagement.ListInviteGroupDefaultOptions)
		agentManagement.PUT("/invite-default-groups/batch", h.AgentManagement.SetInviteGroupDefaultsBatch)
		agentManagement.PUT("/invite-default-groups/:group_id", h.AgentManagement.SetInviteGroupDefault)
		agentManagement.DELETE("/invite-default-groups/:group_id", h.AgentManagement.RemoveInviteGroupDefault)
		agentManagement.GET("/children/:id/groups", h.AgentManagement.ListChildGroupDelegationOptions)
		agentManagement.PUT("/children/:id/groups/batch", h.AgentManagement.SetChildGroupDelegationsBatch)
		agentManagement.PUT("/children/:id/groups/:group_id", h.AgentManagement.SetChildGroupDelegation)
		agentManagement.DELETE("/children/:id/groups/:group_id", h.AgentManagement.RemoveChildGroupDelegation)
	}
}

File: frontend/src/api/agentManagement.ts
import { apiClient } from './client'
import type {
  AgentAllocationSummary,
  AgentAdminTreeResponse,
  AgentChildGroupDelegationOption,
  AgentDirectChildKind,
  AgentDirectChildrenGroupDelegationBatchRequest,
  AgentDirectChildrenGroupDelegationUpdateRequest,
  AgentDirectChildrenGroupDelegationUpdateResponse,
  AgentDirectChildrenGroupQuery,
  AgentDirectChildrenGroupsQuery,
  AgentDirectChildrenGroupDelegationBatchResponse,
  AgentDirectChildrenGroupDelegationReclaimRequest,
  AgentDirectChildrenGroupDelegationReclaimResponse,
  AgentDirectChildrenResponse,
  AgentGroupDelegationBatchRequest,
  AgentGroupDelegationBatchResponse,
  AgentDirectUserCreateRequest,
  AgentGroupDelegationRequest,
  AgentGroupDelegationResponse,
  AgentGroupRate,
  AgentIncomeSetRequest,
  AgentInviteGroupDefaultBatchRequest,
  AgentInviteGroupDefaultBatchResponse,
  AgentInviteGroupDefaultRequest,
  AgentInviteGroupDefaultResponse,
  AgentInviteDefaultsUpdate,
  AgentManagedUser,
  AgentManagementSummary,
  AgentProfile,
  AgentUpgradeRequest,
  AgentAllocationUpdate,
  AgentStructureResponse,
  AgentUsageLog,
  PaginatedResponse
} from '@/types'
import type { AdminUsageQueryParams, AdminUsageStatsResponse } from '@/api/admin/usage'
import type { SimpleApiKey, SimpleUser } from '@/api/admin/usage'

const BASE_PATH = '/agent-management'

export interface AgentUsageAccountSummary {
  id: number
  name: string
}

export interface AgentDirectChildrenQuery {
  search?: string
  page?: number
  page_size?: number
}

const DIRECT_CHILD_KIND_PATH: Record<AgentDirectChildKind, string> = {
  users: 'direct-users',
  agents: 'direct-agents',
  enterprises: 'direct-enterprises',
}

export async function getSummary(): Promise<AgentManagementSummary> {
  const { data } = await apiClient.get<AgentManagementSummary>(`${BASE_PATH}/summary`)
  return data
}

export async function listDirectUsers(query: AgentDirectChildrenQuery = {}): Promise<AgentDirectChildrenResponse> {
  const { data } = await apiClient.get<AgentDirectChildrenResponse>(`${BASE_PATH}/direct-users`, { params: query })
  return data
}

export async function listDirectAgents(query: AgentDirectChildrenQuery = {}): Promise<AgentDirectChildrenResponse> {
  const { data } = await apiClient.get<AgentDirectChildrenResponse>(`${BASE_PATH}/direct-agents`, { params: query })
  return data
}

export async function listDirectEnterprises(query: AgentDirectChildrenQuery = {}): Promise<AgentDirectChildrenResponse> {
  const { data } = await apiClient.get<AgentDirectChildrenResponse>(`${BASE_PATH}/direct-enterprises`, { params: query })
  return data
}

export async function listDirectChildrenWithGroupDelegation(
  kind: AgentDirectChildKind,
  query: AgentDirectChildrenGroupQuery
): Promise<AgentDirectChildrenResponse> {
  const { data } = await apiClient.get<AgentDirectChildrenResponse>(
    `${BASE_PATH}/${DIRECT_CHILD_KIND_PATH[kind]}/groups/assigned`,
    { params: query }
  )
  return data
}

export async function listDirectChildrenWithoutGroupDelegation(
  kind: AgentDirectChildKind,
  query: AgentDirectChildrenGroupsQuery
): Promise<AgentDirectChildrenResponse> {
  const { data } = await apiClient.get<AgentDirectChildrenResponse>(
    `${BASE_PATH}/${DIRECT_CHILD_KIND_PATH[kind]}/groups/unassigned`,
    { params: { ...query, group_ids: query.group_ids.join(',') } }
  )
  return data
}

export async function getAdminAgentTree(): Promise<AgentAdminTreeResponse> {
  const { data } = await apiClient.get<AgentAdminTreeResponse>(`${BASE_PATH}/admin-agent-tree`)
  return data
}

export async function getStructure(ownerId?: number): Promise<AgentStructureResponse> {
  const params = ownerId ? { owner_id: ownerId } : {}
  const { data } = await apiClient.get<AgentStructureResponse>(`${BASE_PATH}/structure`, { params })
  return data
}

export async function listUsage(
  params: AdminUsageQueryParams,
  options?: { signal?: AbortSignal }
): Promise<PaginatedResponse<AgentUsageLog>> {
  const { data } = await apiClient.get<PaginatedResponse<AgentUsageLog>>(`${BASE_PATH}/usage`, {
    params,
    signal: options?.signal
  })
  return data
}

export async function getUsageStats(params: AdminUsageQueryParams): Promise<AdminUsageStatsResponse> {
  const { data } = await apiClient.get<AdminUsageStatsResponse>(`${BASE_PATH}/usage/stats`, { params })
  return data
}

export async function listUsageUsers(): Promise<AgentManagedUser[]> {
  const { data } = await apiClient.get<AgentManagedUser[]>(`${BASE_PATH}/usage/users`)
  return data
}

export async function searchUsageUsers(keyword: string): Promise<SimpleUser[]> {
  const { data } = await apiClient.get<SimpleUser[]>(`${BASE_PATH}/usage/search-users`, {
    params: { q: keyword }
  })
  return data
}

export async function searchUsageApiKeys(userId?: number, keyword?: string): Promise<SimpleApiKey[]> {
  const params: Record<string, unknown> = {}
  if (userId !== undefined && userId !== null) {
    params.user_id = userId
  }
  if (keyword) {
    params.q = keyword
  }
  const { data } = await apiClient.get<SimpleApiKey[]>(`${BASE_PATH}/usage/search-api-keys`, { params })
  return data
}

export async function searchUsageAccounts(keyword: string): Promise<AgentUsageAccountSummary[]> {
  const { data } = await apiClient.get<AgentUsageAccountSummary[]>(`${BASE_PATH}/usage/search-accounts`, {
    params: { q: keyword }
  })
  return data
}

export async function createDirectUser(payload: AgentDirectUserCreateRequest): Promise<AgentManagedUser> {
  const { data } = await apiClient.post<AgentManagedUser>(`${BASE_PATH}/direct-users`, payload)
  return data
}

export async function updateAllocation(childId: number, payload: AgentAllocationUpdate): Promise<AgentAllocationSummary> {
  const { data } = await apiClient.put<AgentAllocationSummary>(`${BASE_PATH}/children/${childId}/allocation`, payload)
  return data
}

export async function updateChildNotes(childId: number, payload: { notes: string }): Promise<AgentManagedUser> {
  const { data } = await apiClient.put<AgentManagedUser>(`${BASE_PATH}/children/${childId}/notes`, payload)
  return data
}

export async function updateInviteDefaults(payload: AgentInviteDefaultsUpdate): Promise<AgentProfile> {
  const { data } = await apiClient.put<AgentProfile>(`${BASE_PATH}/invite-defaults`, payload)
  return data
}

export async function upgradeChild(childId: number, payload: AgentUpgradeRequest): Promise<AgentManagedUser> {
  const { data } = await apiClient.post<AgentManagedUser>(`${BASE_PATH}/children/${childId}/upgrade`, payload)
  return data
}

export async function deleteDirectChild(childId: number): Promise<{ id: number }> {
  const { data } = await apiClient.delete<{ id: number }>(`${BASE_PATH}/children/${childId}`)
  return data
}

export async function listGroups(): Promise<AgentGroupRate[]> {
  const { data } = await apiClient.get<AgentGroupRate[]>(`${BASE_PATH}/groups`)
  return data
}

export async function listChildGroupDelegationOptions(childId: number): Promise<AgentChildGroupDelegationOption[]> {
  const { data } = await apiClient.get<AgentChildGroupDelegationOption[]>(`${BASE_PATH}/children/${childId}/groups`)
  return data
}

export async function listInviteGroupDefaultOptions(): Promise<AgentChildGroupDelegationOption[]> {
  const { data } = await apiClient.get<AgentChildGroupDelegationOption[]>(`${BASE_PATH}/invite-default-groups`)
  return data
}

export async function setChildGroupDelegation(
  childId: number,
  groupId: number,
  payload: AgentGroupDelegationRequest
): Promise<AgentGroupDelegationResponse> {
  const { data } = await apiClient.put<AgentGroupDelegationResponse>(
    `${BASE_PATH}/children/${childId}/groups/${groupId}`,
    payload
  )
  return data
}

export async function setChildGroupDelegationsBatch(
  childId: number,
  payload: AgentGroupDelegationBatchRequest
): Promise<AgentGroupDelegationBatchResponse> {
  const { data } = await apiClient.put<AgentGroupDelegationBatchResponse>(
    `${BASE_PATH}/children/${childId}/groups/batch`,
    payload
  )
  return data
}

export async function setDirectChildrenGroupDelegationsBatch(
  kind: AgentDirectChildKind,
  payload: AgentDirectChildrenGroupDelegationBatchRequest
): Promise<AgentDirectChildrenGroupDelegationBatchResponse> {
  const { data } = await apiClient.put<AgentDirectChildrenGroupDelegationBatchResponse>(
    `${BASE_PATH}/${DIRECT_CHILD_KIND_PATH[kind]}/groups/batch`,
    payload
  )
  return data
}

export async function updateDirectChildrenExistingGroupDelegations(
  kind: AgentDirectChildKind,
  payload: AgentDirectChildrenGroupDelegationUpdateRequest
): Promise<AgentDirectChildrenGroupDelegationUpdateResponse> {
  const { data } = await apiClient.put<AgentDirectChildrenGroupDelegationUpdateResponse>(
    `${BASE_PATH}/${DIRECT_CHILD_KIND_PATH[kind]}/groups/existing`,
    payload
  )
  return data
}

export async function reclaimDirectChildrenGroupDelegations(
  kind: AgentDirectChildKind,
  payload: AgentDirectChildrenGroupDelegationReclaimRequest
): Promise<AgentDirectChildrenGroupDelegationReclaimResponse> {
  const { data } = await apiClient.post<AgentDirectChildrenGroupDelegationReclaimResponse>(
    `${BASE_PATH}/${DIRECT_CHILD_KIND_PATH[kind]}/groups/reclaim`,
    payload
  )
  return data
}

export async function setAgentIncome(childId: number, payload: AgentIncomeSetRequest): Promise<AgentManagedUser> {
  const { data } = await apiClient.put<AgentManagedUser>(`${BASE_PATH}/children/${childId}/agent-income`, payload)
  return data
}

export async function removeChildGroupDelegation(childId: number, groupId: number): Promise<AgentGroupDelegationResponse> {
  const { data } = await apiClient.delete<AgentGroupDelegationResponse>(
    `${BASE_PATH}/children/${childId}/groups/${groupId}`
  )
  return data
}

export async function setInviteGroupDefault(
  groupId: number,
  payload: AgentInviteGroupDefaultRequest
): Promise<AgentInviteGroupDefaultResponse> {
  const { data } = await apiClient.put<AgentInviteGroupDefaultResponse>(
    `${BASE_PATH}/invite-default-groups/${groupId}`,
    payload
  )
  return data
}

export async function setInviteGroupDefaultsBatch(
  payload: AgentInviteGroupDefaultBatchRequest
): Promise<AgentInviteGroupDefaultBatchResponse> {
  const { data } = await apiClient.put<AgentInviteGroupDefaultBatchResponse>(
    `${BASE_PATH}/invite-default-groups/batch`,
    payload
  )
  return data
}

export async function removeInviteGroupDefault(groupId: number): Promise<AgentInviteGroupDefaultResponse> {
  const { data } = await apiClient.delete<AgentInviteGroupDefaultResponse>(
    `${BASE_PATH}/invite-default-groups/${groupId}`
  )
  return data
}

export const agentManagementAPI = {
  getSummary,
  listDirectUsers,
  listDirectAgents,
  listDirectEnterprises,
  listDirectChildrenWithGroupDelegation,
  listDirectChildrenWithoutGroupDelegation,
  getAdminAgentTree,
  getStructure,
  listUsage,
  getUsageStats,
  listUsageUsers,
  searchUsageUsers,
  searchUsageApiKeys,
  searchUsageAccounts,
  createDirectUser,
  updateAllocation,
  updateChildNotes,
  updateInviteDefaults,
  upgradeChild,
  deleteDirectChild,
  listGroups,
  listChildGroupDelegationOptions,
  listInviteGroupDefaultOptions,
  setChildGroupDelegation,
  setChildGroupDelegationsBatch,
  setDirectChildrenGroupDelegationsBatch,
  updateDirectChildrenExistingGroupDelegations,
  reclaimDirectChildrenGroupDelegations,
  setAgentIncome,
  removeChildGroupDelegation,
  setInviteGroupDefault,
  setInviteGroupDefaultsBatch,
  removeInviteGroupDefault,
}

export default agentManagementAPI
